### PR TITLE
Replaced Contig with ContigName in AlignmentRecord and related changes

### PR DESCRIPTION
**Pull request by jpdna**
This PR depends on the related PR #72 [here](https://github.com/bigdatagenomics/bdg-formats/pull/72)   in  `bigdatagenomics/bdg-formats` repo
which supplies  version `0.7.2-SNAPSHOT` of bdg-formats.
With the above available, this PR should compile and pass all tests

See related discussion in #983

The goal of this PR is to remove from `AlignmentRecord` for performance reasons the  `Contig`  and `mateContig` metadata details such as MD5 and URL that are repeated for each contig.  This contig metadata is referred to now only in the `SequenceDictionary` and joined using the contigName and mateContigName which have replaced contig and mateContig in `AlignmentRecord`

Performance Improvement:
Testing on a 3.8 GB BAM file input on a single machine, I see the following improvements based on metrics in Spark web GUI with the changes in this PR as compared to current code as of ` b8e36b2` :
```
27% speed up in MarkDuplicates ( 2.3 minutes to 1.7 minutes )
11% reduction in Shuffle Read and Write size in MarkDuplicates ( 10.1GB to 9.0 GB)
```

### CODE REVIEW

File: adam-apis/pom.xml [hunk 1]
```diff
@@ -106,6 +106,7 @@
     <dependency>
       <groupId>org.bdgenomics.bdg-formats</groupId>
       <artifactId>bdg-formats</artifactId>
+      <version>0.7.2-SNAPSHOT</version>
```
Comment (fnothaft) [inline adam-apis/pom.xml]: You don't need the version here, it inherits from the `${ADAM_HOME}/pom.xml`.

File: adam-cli/pom.xml [hunk 2]
```diff
@@ -167,6 +167,7 @@
     <dependency>
       <groupId>org.bdgenomics.bdg-formats</groupId>
       <artifactId>bdg-formats</artifactId>
+      <version>0.7.2-SNAPSHOT</version>
```
Comment (fnothaft) [inline adam-cli/pom.xml]: Ditto RE: above.

File: adam-core/pom.xml [hunk 3]
```diff
@@ -118,6 +118,7 @@
     <dependency>
       <groupId>org.bdgenomics.bdg-formats</groupId>
       <artifactId>bdg-formats</artifactId>
+      <version>0.7.2-SNAPSHOT</version>
```
Comment (fnothaft) [inline adam-core/pom.xml]: Ditto.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/AlignmentRecordConverter.scala [hunk 4]
```diff
@@ -155,8 +154,8 @@ class AlignmentRecordConverter extends Serializable {
         // only set alignment flags if read is aligned
         if (m) {
           // if we are aligned, we must have a reference
-          assert(adamRecord.getContig != null, "Cannot have null contig if aligned.")
-          builder.setReferenceName(adamRecord.getContig.getContigName)
+          assert(adamRecord.getContigName != null, "Cannot have null contig if aligned.")
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/AlignmentRecordConverter.scala]: Small nit: while we're changing this line, we should probably `s/assert/require/g`.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala [hunk 5]
```diff
@@ -81,7 +81,7 @@ class SAMRecordConverter extends Serializable with Logging {
       val readReference: Int = samRecord.getReferenceIndex
       if (readReference != SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
         dict(samRecord.getReferenceName).foreach { (rec) =>
-          builder.setContig(SequenceRecord.toADAMContig(rec))
+          builder.setContigName(SequenceRecord.toADAMContig(rec).getContigName)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala]: I think you can remove the `.foreach` and just simplify to:

```
builder.setContigName(samRecord.getReferenceName)
```

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala [hunk 6]
```diff
@@ -128,7 +128,7 @@ class SAMRecordConverter extends Serializable with Logging {
 
       if (mateReference != SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX) {
         dict(samRecord.getMateReferenceName).foreach { (rec) =>
-          builder.setMateContig(SequenceRecord.toADAMContig(rec))
+          builder.setMateContigName(SequenceRecord.toADAMContig(rec).getContigName)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/SAMRecordConverter.scala]: Ditto here RE: `.foreach`.

File: adam-core/src/main/scala/org/bdgenomics/adam/util/Util.scala [hunk 7]
```diff
@@ -16,9 +16,13 @@
  * limitations under the License.
  */
 package org.bdgenomics.adam.util
-
 import org.bdgenomics.formats.avro.Contig
 
+// We may want to reintroduce in the future MD5 concordance test for isSameContig as
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/Util.scala]: I don't like to leave dead code in. Let's either remove the whole comment + code, or keep the comment but include a commit hash for when the code was removed so that it is easy to track down the diff with the old code.

As an aside, I would remove this class. IMO, it's not very useful to have a function that just calls `String.equals`.

File: pom.xml [hunk 8]
```diff
@@ -316,7 +316,7 @@
       <dependency>
         <groupId>org.bdgenomics.bdg-formats</groupId>
         <artifactId>bdg-formats</artifactId>
-        <version>0.7.0</version>
+        <version>0.7.2-SNAPSHOT</version>
```
Comment (fnothaft) [inline pom.xml]: We'll cut a new release of `bdg-formats` to merge this.

**Comment by fnothaft** [comment]
Aside from a few nits, this LGTM! Thanks for profiling the change, @jpdna!

**Comment by jpdna** [comment]
Made the changes suggested by @fnothaft above and squashed

File: pom.xml (hunk 8 above)
Comment (fnothaft) [inline pom.xml]: I am cutting the 0.7.1 bdg-formats release right now.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala [hunk 9]
```diff
@@ -21,18 +21,14 @@ import org.apache.spark.rdd.RDD
 import org.bdgenomics.adam.converters.AlignmentRecordConverter
 import org.bdgenomics.adam.models.SequenceRecord
 import org.bdgenomics.adam.rdd.ADAMSequenceDictionaryRDDAggregator
+import org.apache.spark.Logging
 import org.bdgenomics.formats.avro._
 import scala.collection.JavaConversions._
 
-class FragmentRDDFunctions(rdd: RDD[Fragment]) extends ADAMSequenceDictionaryRDDAggregator[Fragment](rdd) {
+class FragmentRDDFunctions(rdd: RDD[Fragment]) extends Serializable with Logging {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala]: Why is this removed?  I'm actually not sure what ADAMSequenceDictionaryRDDAggregator was for in the first place.

File: adam-core/src/test/scala/org/bdgenomics/adam/util/UtilSuite.scala [hunk 10]
```diff
@@ -24,20 +24,23 @@ class UtilSuite extends ADAMFunSuite {
...
 
     // proper null handling
     assert(Util.isSameContig(null, null))
-    assert(!Util.isSameContig(null, b.build()))
-    assert(!Util.isSameContig(a.build(), null))
+    assert(!Util.isSameContig(null, b.build().getContigName))
+    assert(!Util.isSameContig(a.build().getContigName, null))
 
-    a.setContigMD5("md5")
+    // Prior to removing Contig from AlignmentRecprd the code below
```
Comment (heuermh) [inline adam-core/src/test/scala/org/bdgenomics/adam/util/UtilSuite.scala]: minor typo

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala (hunk 9 above)
Comment (jpdna) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala]: > ADAMSequenceDictionaryRDDAggregator
seemed to be used to extract SeqDicts from existing object, but didn't seem to make sense anymore in context of having removed Contig from AlignmentRecord, including within Fragmen - and importantly was not actually used in any code or tests.  Extending ADAMSequenceDictionaryRDDAggregator here in fact blocked the Contig factoring out as it requires data which is no longer in AlignmentRecord.  I searched for any usages of the removed functions  and there were none, so I removed it as a superclass.

Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala]: Sounds good to me, I'm always in favor of removing unnecessary code.

I wonder if this changes anyone's opinion on this commit https://github.com/ryan-williams/adam/commit/c5a8f51f7d1bc38cc80f9f0699b5ac92947ecf31 that adds extension to some of the Functions classes.

File: pom.xml (hunk 8 above)
Comment (fnothaft) [inline pom.xml]: [0.7.1 has propegated to Maven Central](http://search.maven.org/#artifactdetails%7Corg.bdgenomics.bdg-formats%7Cbdg-formats%7C0.7.1%7Cjar).

Comment (fnothaft) [inline pom.xml]: [0.7.1 has propagated to Maven Central](http://search.maven.org/#artifactdetails%7Corg.bdgenomics.bdg-formats%7Cbdg-formats%7C0.7.1%7Cjar).

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala (hunk 9 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDDFunctions.scala]: Both this and the commit you pointed at @heuermh LGTM.

**Comment by jpdna** [comment]
bdg-formats dependency has been updated to 0.7.1 and change squashed,  so I think this is ready to go now

**Comment by jpdna** [comment]
well crud, github webpage on my fork is not updating with my push, though git thinks I pushed.  Github was down a few minutes ago - I'll ping this thread again in when it matches.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
LGTM now. Let's discuss on the call tomorrow and make sure everyone is good with merging.

**Comment by tdanford** [comment]
+1000 from me

**Comment by fnothaft** [comment]
Thanks @jpdna! Merged.
